feature/ja-61_RepositoryBase

skrawus opened the pull request.

skrawus committed:
Add RepositoryBase

Zjyslav commented on TutorLizard.BusinessLogic/Data/RepositoryBase.cs:
```diff
@@ -0,0 +1,71 @@
+                fullPath = Path.Combine(AppContext.BaseDirectory, filePath);
+
+            if (!File.Exists(fullPath))
+                _data = new List<T>();
```
Replace this line (35) with `return;`
We changed return to assignment, but it's actually important to return here (but the value is not important).
`_data` initializes itself, so there's no need to assign empty list to it.

skrawus replied:
This forced me to make few more changes (so the code looks like in DA now)

Zjyslav replied:
If you did `return;` instead of `return new List<T>();` you wouldn't have to make these changes.

Zjyslav reviewed (changes requested):
I looked into the way we could implement getting `filePath` from configuration and if we want to do it in the derived classes, specifying constructor in the base class hurts us, because it's impossible to set the `_filePath` in derived constructor before the base constructor calls `LoadFromJson()` that uses this path.

I think that the most flexible approach would be to remove the constructor and replace:
`private string _filePath;`
with
`protected abstract string FilePath { get; }`

This will force any derived class to implement `FilePath`, but allow to do it either specify the value inline with the implementation or in the constructor. Each implementation would  have to call `LoadFromJson()`, but it would have the freedom to decide when it's ready to to so.
Implementation could look like this:
```
public class AdRepository : RepositoryBase<Ad>
{
    protected override string FilePath => "Data/Ads.json";
    public AdRepository()
    {
        LoadFromJson();
    }
}
```

or: 
```
public class UserRepository : RepositoryBase<User>
{
    protected override string FilePath { get; }

    public UserRepository(IConfiguration config)
    {
        string section = "DataFilePaths:Users";
        string? path = config.GetValue<string>(section);
        if (string.IsNullOrWhiteSpace(path))
            throw new Exception($"Configuration section missing: {section}");
        FilePath = path;
        LoadFromJson();
    }
}
```

skrawus committed:
Fix LoadFromJson

Add return in line 35

skrawus committed:
Delete constructor

skrawus committed:
Fix LoadFromJson

If we want to return here, it cannot be void. Also some other things had to be changed.

skrawus committed:
Create property FilePath

Create FilePath and fix code where the FilePath is used

skrawus commented:
The requested changes are pushed. :)

skrawus committed:
Fix LoadFromJson

And add 'return'
